From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Etaash Mathamsetty
 <45927311+Etaash-mathamsetty@users.noreply.github.com>
Date: Mon, 30 Mar 2026 01:03:35 -0500
Subject: [PATCH 04/46] 
 0008-winewayland-respect-size-hint-when-restoring-from-ti.patch

---
 dlls/winewayland.drv/window.c      | 10 ++++++++--
 dlls/winewayland.drv/window.c.orig | 16 ++++++++++++++++
 2 files changed, 24 insertions(+), 2 deletions(-)

diff --git a/dlls/winewayland.drv/window.c b/dlls/winewayland.drv/window.c
index 813f809d30b..0ebad1df465 100644
--- a/dlls/winewayland.drv/window.c
+++ b/dlls/winewayland.drv/window.c
@@ -531,8 +531,14 @@ static void wayland_configure_window(HWND hwnd)
 
     state = surface->processing.state;
     /* Ignore size hints if we don't have a state that requires strict
-     * size adherence, in order to avoid spurious resizes. */
-    if (state)
+     * size adherence, in order to avoid spurious resizes.
+     * The tiled and maximized states have a strict size adherance, so
+     * their sizes cannot change while we are still processing the new config.
+     * This allows us to respect the size hint on transitions of maximized/tiled
+     * to a stateless regular window. */
+    if (state || surface->current.state &
+        (WAYLAND_SURFACE_CONFIG_STATE_MAXIMIZED |
+               WAYLAND_SURFACE_CONFIG_STATE_TILED))
     {
         width = surface->processing.rect.right - surface->processing.rect.left;
         height = surface->processing.rect.bottom - surface->processing.rect.top;
diff --git a/dlls/winewayland.drv/window.c.orig b/dlls/winewayland.drv/window.c.orig
index afef743755b..813f809d30b 100644
--- a/dlls/winewayland.drv/window.c.orig
+++ b/dlls/winewayland.drv/window.c.orig
@@ -827,6 +827,22 @@ void WAYLAND_UpdateLayeredWindow(HWND hwnd, BYTE alpha, UINT flags)
     wayland_win_data_release(data);
 }
 
+/***********************************************************************
+ *          WAYLAND_HasWindowManager
+ */
+BOOL WAYLAND_HasWindowManager(const char *name)
+{
+    static int once;
+    const char *env = getenv("XDG_CURRENT_DESKTOP");
+
+    if (!once++) TRACE("DE: %s\n", debugstr_a(env));
+
+    if (!strcmp("waylanddrv", name)) return TRUE;
+    if (env && !strcmp(env, name)) return TRUE;
+
+    return FALSE;
+}
+
 void set_client_surface(HWND hwnd, struct wayland_client_surface *new_client)
 {
     HWND toplevel = NtUserGetAncestor(hwnd, GA_ROOT);
